macros/src/lib.rs

```diff
@@ -0,0 +1,517 @@
+                    #[export_name = #name_s]
+                    #(#attrs)*
+                    pub fn #ident() {
+                        extern crate cortex_m_rt;
+
+                        cortex_m_rt::Exception::#name;
+
+                        static mut __STATE__: #ty = #expr;
+
+                        #[allow(non_snake_case)]
+                        let #state: &mut #ty = unsafe { &mut __STATE__ };
+
+                        #(#stmts)*
+                    }
+                ).into()
+            } else {
+                quote!(
+                    #[export_name = #name_s]
+                    #(#attrs)*
+                    pub fn #ident() {
+                        extern crate cortex_m_rt;
+
+                        cortex_m_rt::Exception::#name;
+
+                        #(#stmts)*
+                    }
+                ).into()
+            }
+        }
+    }
+}
+
+/// Attribute to mark which function will be called at the beginning of the reset handler.
+///
+/// The function must have the signature of `unsafe fn()`.
+///
+/// The function passed will be called before static variables are initialized. Any access of static
+/// variables will result in undefined behavior.
+///
+/// # Examples
+///
+/// ```
+/// # use cortex_m_rt_macros::pre_init;
+/// #[pre_init]
+/// unsafe fn before_main() {
+///     // do something here
+/// }
+///
+/// # fn main() {}
+/// ```
+#[proc_macro_attribute]
+pub fn pre_init(args: TokenStream, input: TokenStream) -> TokenStream {
+    let f: ItemFn = syn::parse(input).expect("`#[pre_init]` must be applied to a function");
+
+    // check the function signature
+    assert!(
+        f.constness.is_none()
+            && f.vis == Visibility::Inherited
+            && f.unsafety.is_some()
+            && f.abi.is_none()
+            && f.decl.inputs.is_empty()
+            && f.decl.generics.params.is_empty()
+            && f.decl.generics.where_clause.is_none()
+            && f.decl.variadic.is_none()
+            && match f.decl.output {
+                ReturnType::Default => true,
+                ReturnType::Type(_, ref ty) => match **ty {
+                    Type::Tuple(ref tuple) => tuple.elems.is_empty(),
+                    _ => false,
+                },
+            },
+        "`#[pre_init]` function must have signature `unsafe fn()`"
+    );
+
+    assert_eq!(
+        args.to_string(),
+        "",
+        "`pre_init` attribute must have no arguments"
+    );
+
+    // XXX should we blacklist other attributes?
+    let attrs = f.attrs;
+    let ident = f.ident;
+    let block = f.block;
+
+    quote!(
+        #[export_name = "__pre_init"]
+        #(#attrs)*
+        pub unsafe fn #ident() #block
+    ).into()
+}
+
+/// Attribute to place a function in RAM
+///
+/// This attribute can *not* be used on generic functions.
+///
+/// # Examples
+///
+/// ```
+/// # use cortex_m_rt_macros::ramfunc;
+/// #[ramfunc]
+/// unsafe fn computation_heavy_function() {
```

Comment:
Why `unsafe`? Also I think it would be better do a *computation heavy* function example with something that actually consumes and/or produces something via parameters and/or return values. ;) Showing how to apply this to an interrupt/exception handler would also be helpful.

Reply:
> Why unsafe?

Oh, I literally copy pasted the `#[pre_init]` example and forgot to remove the unsafe keyword. Will fix it in a bit.